server/sonar-web/src/main/js/components/workspace/Workspace.tsx

```diff
@@ -0,0 +1,235 @@
+/*
+ * SonarQube
+ * Copyright (C) 2009-2018 SonarSource SA
+ * mailto:info AT sonarsource DOT com
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 3 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this program; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+import * as React from 'react';
+import * as PropTypes from 'prop-types';
+import { omit, uniqBy } from 'lodash';
+import { WorkspaceContext, ComponentDescriptor, RuleDescriptor } from './context';
+import WorkspaceNav from './WorkspaceNav';
+import WorkspacePortal from './WorkspacePortal';
+import { lazyLoad } from '../lazyLoad';
+import './styles.css';
+
+const WorkspaceRuleViewer = lazyLoad(() => import('./WorkspaceRuleViewer'));
+const WorkspaceComponentViewer = lazyLoad(() => import('./WorkspaceComponentViewer'));
+
+interface State {
+  components: ComponentDescriptor[];
+  height: number;
+  maximized?: boolean;
+  open: { component?: string; rule?: string };
+  rules: RuleDescriptor[];
+}
+
+const MIN_HEIGHT = 0.05;
+const MAX_HEIGHT = 0.85;
+const INITIAL_HEIGHT = 300;
+
+const STORAGE_KEY = 'sonarqube-workspace';
+const TYPE_KEY = '__type__';
+
+export default class Workspace extends React.PureComponent<{}, State> {
+  mounted = false;
+
+  static childContextTypes = {
+    workspace: PropTypes.object
+  };
+
+  constructor(props: {}) {
+    super(props);
+    this.state = { height: INITIAL_HEIGHT, open: {}, ...this.loadWorkspace() };
+  }
+
+  getChildContext = (): { workspace: WorkspaceContext } => {
+    return { workspace: { openComponent: this.openComponent, openRule: this.openRule } };
+  };
+
+  componentDidMount() {
+    this.mounted = true;
+  }
+
+  componentDidUpdate(_: {}, prevState: State) {
+    if (prevState.components !== this.state.components || prevState.rules !== this.state.rules) {
+      this.saveWorkspace();
+    }
+  }
+
+  componentWillUnmount() {
+    this.mounted = false;
+  }
+
+  loadWorkspace = () => {
+    try {
+      const data: any[] = JSON.parse(window.localStorage.getItem(STORAGE_KEY) || '');
+      const components: ComponentDescriptor[] = data.filter(x => x[TYPE_KEY] === 'component');
+      const rules: RuleDescriptor[] = data.filter(x => x[TYPE_KEY] === 'rule');
+      return { components, rules };
+    } catch {
+      // fail silently
+      return { components: [], rules: [] };
+    }
+  };
+
+  saveWorkspace = () => {
```

Comment:
I thought about that, but decided to keep in here, because I don't want to pass `ComponentDescriptor` and `RuleDescriptor` types and `TYPE_KEY` constant. This is all related only to workspace.

Reply:
Ok, maybe only the part about accessing the storage and the key then ? So we only have one place from where we access the localStorage ?